### PR TITLE
Fix for the errors i ran into in Issue4344

### PR DESCRIPTION
**Pull request by smurfd**
Adding check if not in CI for GC tests
This way i can run: python3 -m pytest test/ on mac without any errors.
And also be able to use python3.8

### CODE REVIEW

File: test/test_gc.py [hunk 1]
```diff
@@ -17,6 +19,7 @@ def test_gc(self):
     del a,b
     assert(tensors_allocated() == 1) # one for Tensor._rng_counter
 
+  @unittest.skipIf(not CI, "need multiprocessing to work")
```
Comment (geohot) [inline test/test_gc.py:22]: I don't understand why this should fail not in CI. In general, GC failure is caused by another test failing and not cleaning up.

File: test/test_linearizer_failures.py [hunk 2]
```diff
@@ -236,6 +236,8 @@ def test_failure_31(self):
     opts = [Opt(op=OptOps.UNROLL, axis=0, amt=0), Opt(op=OptOps.PADTO, axis=1, amt=32)]
     helper_test_lin(Linearizer(ast), opts=opts, failed_platforms=[])
 
+  # AssertionError: failed on METAL with COMPARE_ERROR
+  @unittest.skipIf(Device.DEFAULT=="METAL", "flaky on METAL")
```
Comment (geohot) [inline test/test_linearizer_failures.py:240]: This only fails on local Metal, not CI Metal?

**Comment by geohot** [comment]
Would merge the typing thing standalone

**Comment by smurfd** [comment]
sure, splitting the typing stuff to a separate commit.

File: test/test_linearizer_failures.py (hunk 2 above)
Comment (smurfd) [reply test/test_linearizer_failures.py:240]: yes, only local for me atleast.

File: test/test_gc.py (hunk 1 above)
Comment (smurfd) [reply test/test_gc.py:22]: I will try to show. (with everything else applied, and not skipping gc tests)
This is local. The problem seems related to the test suite not being run with number of workers >=2, (which you do in CI)
And running only `test/test_gc.py` of course works.

The reason i chose `skipIf(not CI` is that i originally wanted to check if the tests were run with `-n` but that was, afaik, not so nice and easy to do.

Doing some bisecting, i tried to move around(renaming) `test_gc.py` and it seems the problem is in `test/models/test_end2end.py` (of course verify)
Meaning, i rename it to `test/models/test_e1gc.py` so it runs before `test/models/test_efficientnet.py` gc test pass. renaming it to `test/models/test_exgc.py` gc test fails.

If i commenting away:
https://github.com/tinygrad/tinygrad/blob/828c98d5c41e6b3cda97475f3c6910e184f07116/test/models/test_end2end.py#L67-L69
ofcourse fails that test, but `test_gc.py` passes

without touching `test/models/test_end2end.py`
failing (its only the gc tests that shows as failed)
```
$ python3 -m pytest test/     
========================================================== test session starts ===========================================================
platform darwin -- Python 3.8.19, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/willygorillatg/tinygrad
plugins: hypothesis-6.103.1, xdist-3.6.1
collected 1782 items                           
```
.....
```
======================================================== short test summary info =========================================================
FAILED test/test_gc.py::TestGC::test_gc - ReferenceError: weakly-referenced object no longer exists
FAILED test/test_gc.py::TestGC::test_gc_complex - ReferenceError: weakly-referenced object no longer exists
=========================== 2 failed, 1567 passed, 177 skipped, 36 xfailed, 192 warnings in 141.32s (0:02:21) ======
```

ok
```
$ python3 -m pytest -n=2 test/
========================================================== test session starts ===========================================================
platform darwin -- Python 3.8.19, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/willygorillatg/tinygrad
plugins: hypothesis-6.103.1, xdist-3.6.1
2 workers [1782 items]  
```
.....
```
================================= 1569 passed, 177 skipped, 36 xfailed, 222 warnings in 95.87s (0:01:35) ===
```

File: test/test_linearizer_failures.py (hunk 2 above)
Comment (Qazalin) [reply test/test_linearizer_failures.py:240]: I can't repro, Metal 3 chip M2 max

File: test/test_gc.py (hunk 1 above)
Comment (Qazalin) [reply test/test_gc.py:22]: yes, without -n= it will fail. But `not CI` doesn't look like the right fix.

Comment (smurfd) [reply test/test_gc.py:22]: the easiest fix, is to add `-n=auto` to the README.md 
```
python3 -m pytest test/                 # whole test suite
```
to
```
python3 -m pytest -n=auto test/                 # whole test suite
```

File: test/test_linearizer_failures.py (hunk 2 above)
Comment (smurfd) [reply test/test_linearizer_failures.py:240]: not sure metal version, but m3 pro
if its only for me, then its probably ok to close
freshly cloned repo 898430c004df1fa97a1c613ffb356a09b8175abe

```
$ python3 -m pytest test/test_linearizer_failures.py
========================================================== test session starts ===========================================================
platform darwin -- Python 3.8.19, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/willygorillatg/tinygrad
plugins: hypothesis-6.103.1, xdist-3.6.1
collected 31 items                                                                                                                       

test/test_linearizer_failures.py ..............s..........F.....                                                                   [100%]

================================================================ FAILURES ================================================================
_________________________________________________ TestLinearizerFailures.test_failure_32 _________________________________________________

self = <test.test_linearizer_failures.TestLinearizerFailures testMethod=test_failure_32>

    @unittest.skipIf(CI, "for real AMD GPU")
    def test_failure_32(self):
      # kernel from beaming resnet
      # Memory access fault on tinybox red
      ast = LazyOp(op=BufferOps.STORE, src=(LazyOp(op=UnaryOps.CAST, src=(LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=UnaryOps.CAST, src=(LazyOp(op=BinaryOps.MUL, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.half, st=ShapeTracker(views=(View(shape=(1, 256, 1, 256, 4, 16, 4, 16), strides=(0, 50176, 0, 196, 0, 14, 0, 1), offset=-15, mask=((0, 1), (0, 256), (0, 1), (0, 256), (0, 4), (1, 15), (0, 4), (1, 15)), contiguous=False), View(shape=(256, 1, 256, 14, 14, 256, 3, 3), strides=(1048576, 0, 0, 64, 1, 4096, 1088, 17), offset=0, mask=None, contiguous=False))))), LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=2, dtype=dtypes.half, st=ShapeTracker(views=(View(shape=(256, 1, 256, 14, 14, 256, 3, 3), strides=(0, 0, 2304, 0, 0, 9, 3, 1), offset=0, mask=None, contiguous=False),))))), arg=None),), arg=dtypes.float),), arg=(7, 6, 5)),), arg=dtypes.half),), arg=MemBuffer(idx=0, dtype=dtypes.half, st=ShapeTracker(views=(View(shape=(256, 1, 256, 14, 14, 1, 1, 1), strides=(50176, 0, 196, 14, 1, 0, 0, 0), offset=0, mask=None, contiguous=True),))))
      opts = [Opt(op=OptOps.TC, axis=2, amt=2), Opt(op=OptOps.UPCAST, axis=2, amt=7), Opt(op=OptOps.UNROLL, axis=1, amt=0), Opt(op=OptOps.LOCAL, axis=1, amt=16)]
>     helper_test_lin(Linearizer(ast), opts=opts, failed_platforms=[])

test/test_linearizer_failures.py:245: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

lin = <tinygrad.codegen.linearizer.Linearizer object at 0x137fecd30>
opts = [Opt(op=OptOps.TC, axis=2, amt=2), Opt(op=OptOps.UPCAST, axis=2, amt=7), Opt(op=OptOps.UNROLL, axis=1, amt=0), Opt(op=OptOps.LOCAL, axis=1, amt=16)]
failed_platforms = [], rtol = 0.01, atol = 0.01

    def helper_test_lin(lin: Linearizer, opts, failed_platforms, rtol=1e-2, atol=1e-2):
      if any(b.dtype == dtypes.half for b in lin.membufs) and not is_dtype_supported(dtypes.half): return
      if any(b.dtype == dtypes.bfloat16 for b in lin.membufs) and not is_dtype_supported(dtypes.bfloat16): return
    
      for opt in opts:
        try:
          lin.apply_opt(opt)
        except KernelOptError:
          # it's considered fixed if we invalidated the opts
          assert Device.DEFAULT not in failed_platforms, f"unexpected success on {Device.DEFAULT}"
          return
    
      compare_result = compare_linearizer(lin, rtol=rtol, atol=atol)
      if compare_result[0] in ["PASS", "KernelOptError"]:
        # it's considered fixed if we invalidated the opts
        assert Device.DEFAULT not in failed_platforms, f"unexpected success on {Device.DEFAULT}"
      else:
>       assert Device.DEFAULT in failed_platforms, f"failed on {Device.DEFAULT} with {compare_result[0]}"
E       AssertionError: failed on METAL with COMPARE_ERROR
E       assert 'METAL' in []
E        +  where 'METAL' = <tinygrad.device._Device object at 0x1070eab80>.DEFAULT

test/test_linearizer_failures.py:33: AssertionError
======================================================== short test summary info =========================================================
FAILED test/test_linearizer_failures.py::TestLinearizerFailures::test_failure_32 - AssertionError: failed on METAL with COMPARE_ERROR
================================================ 1 failed, 29 passed, 1 skipped in 7.66s =================================================
```

**Comment by smurfd** [comment]
ok, lets do like this :)
Im not sure i follow why running without `-n=auto` would fail to garbage collect, and the actual test that is not behaving correctly is not showing the error...

And i guess `test/test_linearizer_failures.py` was only me...

Atleast now the README.md is corrrect, will make it easier for people to test locally, before submitting a patch.

**Comment by geohot** [comment]
Often when I'm running tests locally I don't want them in parallel. If people do, it's just generic pytest. Won't merge.

**Pull request closed without merging**